mods/default/functions.lua

```diff
@@ -576,3 +576,118 @@ function default.can_interact_with_node(player, pos)
 
 	return false
 end
+
+
+--
+-- Stone/Ore API
+--
+
+default.registered_stones = {}
+default.registered_ores = {}
+
+local function check_stone_groups(groups_a, groups_b)
+	for fieldname, value in pairs(groups_a) do
+		if groups_b[fieldname] == value then
+			return true
+		end
+	end
+	return false
+end
+
+local function register_stone_with_ore(ore, stone)
+	-- set name, description, texture and groups
+	-- name format: 'modNameStone:stoneName_with_oreName'
+	-- for example: 'morestones:granite_with_iron'
+	local fullname = stone.name .. "_with_" .. ore.name
+
+	-- description format: 'Stonename Orename Ore' -> 'Granite Iron Ore'
+	ore.description = stone.description .. " " .. ore.description .. " Ore"
+	-- take the stone texture and add the ore texture as overlay
+	ore.tiles = {stone.tiles[1] .. "^" .. ore.texture}
+	-- make the ore as cracky as the stone
+	ore.groups.cracky = (stone.groups.cracky or 3) + ore.additional_stone_crackyness or 0
+	if ore.groups.cracky < 1 then
+		ore.groups.cracky = 1
+	end
+
+	-- use sounds from stone
+	ore.sounds = stone.sounds
+
+	-- clean up
+	local ore_mapgen = ore.mapgen or {}
+	ore.additional_stone_crackyness = nil
+	ore.mapgen = nil
+	ore.stone_groups = nil
+	ore.texture = nil
+	ore.name = nil
+
+	minetest.register_node(fullname, ore)
+
+	-- return if no map generations to register
+	if #ore_mapgen == 0 then return end
+
+	-- register ore generations
+	for _, oredef in pairs(ore_mapgen) do
+		if oredef.stone_groups then
+			if not check_stone_groups(stone.stone_groups, oredef.stone_groups) then
+				break
+			end
+		end
+
+		-- set the fitting stone name
+		oredef.wherein = stone.name
+		-- default to use scatter ores
+		oredef.ore_type = oredef.ore_type or "scatter"
+		-- set ore itemstring
+		oredef.ore = fullname
+
+		minetest.register_ore(oredef)
+	end
+end
+
+function default.register_stone(def)
+	-- make sure that these exist
+	def.name = def.name or "stone"
+	def.modname = def.modname or core.get_current_modname()
+	def.description = def.description or "Stone"
+	def.tiles = def.tiles or {"unknown_node.png"}
+	def.groups = def.groups or {}
+	def.groups.cracky = def.groups.cracky or 3
+	def.stone_groups = def.stone_groups or {}
+	def.stone_groups.all = 1 -- for ores in all stones
+	def.sounds = def.sounds or default.node_sound_stone_defaults()
```

Comment:
These few ores are only registered once and a function call takes almost no time, so this can entirely be ignored.